lib/sensors/src/URM04Sensor.cpp

```diff
@@ -0,0 +1,188 @@
+    // error occured in trigger step
+    return false;
+  } else {
+    // no error occured in trigger step
+    // flush the buffer from serial
+    m_serial.sync();
+    // wait for at least 30 ms after calling trigger function
+    ThisThread::sleep_for(35ms);
+    return true;
+  }
+}
+
+// pass by reference a variable to hold the distance value - will give MAX_FLOAT if out of range
+bool URM04Sensor::URM04Sensor::read_distance(float& distance) {
+  /****************** TRIGGER SENSOR BEFORE READING DISTANCE ********************/
+  // if trigger fails set distance to -1 and return false
+  if (trigger_sensor() == false) {
+    distance = -1;
+    return false;
+  }
+
+  /************ INSTANTIATE COMMANDS TO BE SENT OVER SERIAL************/
+  // check sum represents the final bit in command buffer - made by adding all previous bits in command buffer
+  uint8_t checkSum;
+  // buffer header
+  m_cmdst[0] = 0x55;
+  m_cmdst[1] = 0xAA;
+  // device address
+  m_cmdst[2] = m_startAddr;
+  // command length
+  m_cmdst[3] = 0x00;
+  // the command itself
+  m_cmdst[4] = 0x02;
+  // checksum bit
+  checkSum = m_cmdst[0] + m_cmdst[1] + m_cmdst[2] + m_cmdst[3] + m_cmdst[4];
+  // instantiate the last element in the command buffer with checksum value
+  m_cmdst[5] = checkSum;
+
+  /************** SEND COMMANDS OVER SERIAL***********/
+  // returns the number of bytes if successful read
+  int w_num_bytes = m_serial.write(&m_cmdst[0], 6);
+
+  // else if numb_btyes < 0 serial failed or if number of bytes read != 6 (because 6 bytes sent over serial)
+  if (w_num_bytes != 6) {
+    // return false indicating read command has failed
+    distance = -1;
+    return false;
+  }
+
+  // flush buffer from serial
+  m_serial.sync();
+
+  /******* READ RETURN VALUE FROM SERIAL**************/
+
+  // turn on reading mode for RS485
+  m_trigPin.write(LOW);
+  ThisThread::sleep_for(1ms);
+
+  // returns the number of bytes if successful read
+  int r_num_bytes = m_serial.read(&m_cmdst[0], 8);
+
+  // else if numb_btyes < 0 serial failed or if number of bytes read != 8 (because 8 bytes sent over serial)
+  if (r_num_bytes != 8) {
+    distance = -1;
+    return false;
+  }
+  /******** PARSE THROUGH THE DATA READ FROM SERIAL*********/
+  else {
+    // check if the checksum is incorrect
+    if (m_cmdst[7] != m_cmdst[0] + m_cmdst[1] + m_cmdst[2] + m_cmdst[3] + m_cmdst[4] + m_cmdst[5] + m_cmdst[6]) {
+      // return false if checksum failed
+      distance = -1;
+      return false;
+    } else {
+      // get distance from sensor
+      distance = (float)(m_cmdst[5] << 8) + (float)m_cmdst[6];
+      // --------------------- or --------------------------
+      // distance = (float)(m_cmdst[5]*256) + (float)m_cmdst[6];
+
+      // check if distance is out of range - if so return maximum float value
+      if (m_cmdst[5] == 0xFF && m_cmdst[6] == 0xFF) {
+        distance = MAX_FLOAT;
+      }
+      return true;
+    }
+  }
+  // flush serial
+  m_serial.sync();
```

Comment:
This line will never be executed since the function will always return before getting here

Reply:
Alright.